pkg/apis/scaler/v1alpha2/crd_demand.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2019 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha2
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// DemandPhaseEmpty is the state of a demand object when it is first created
+	DemandPhaseEmpty string = ""
+	// DemandPhasePending is the state a demand object is in when Scaler has acknowledged it but has not yet taken
+	// action to fulfill the demand
+	DemandPhasePending string = "pending"
+	// DemandPhaseFulfilled is the state a demand object is in when Scaler has taken action and the action has completed
+	// to fulfill the demand. At this point, it is expected that there is capacity to meet the demand the object represents
+	DemandPhaseFulfilled string = "fulfilled"
+	// DemandPhaseCannotFulfill is the state a demand object is in when Scaler is unable to satisfy the demand. This is
+	// possible if the demand contains a single unit that is larger than the instance group is configured to use, or if
+	// the instance group has reached its maximum capacity and cannot allocate more
+	DemandPhaseCannotFulfill string = "cannot-fulfill"
+
+	// ResourceCPU is the name of CPU resource.
+	ResourceCPU corev1.ResourceName = corev1.ResourceCPU
+	// ResourceMemory is the name of Memory resource.
+	ResourceMemory corev1.ResourceName = corev1.ResourceMemory
+	// ResourceNvidiaGPU is the name of Nvidia GPU resource.
+	ResourceNvidiaGPU corev1.ResourceName = "nvidia.com/gpu"
+)
+
+var (
+	// AllDemandPhases is a list of all phases that a demand object could be in
+	AllDemandPhases = []string{
+		DemandPhaseEmpty,
+		DemandPhasePending,
+		DemandPhaseFulfilled,
+		DemandPhaseCannotFulfill,
+	}
+
+	// AllSupportedResources is a list of all resources that the demand object supports
+	AllSupportedResources = []corev1.ResourceName{
```

Comment:
Why do we need this? We should be able to support generic resources.

Reply:
We should define what resources are supported first class by spark scheduler. 

With out this users won't know that `nvidia.com/gpu` is supported by spark scheduler
With out this users won't know that `glob-resource` is not supported spark scheduler